utils/src/test/java/software/amazon/awssdk/utils/cache/lru/LruCacheTest.java

```diff
@@ -0,0 +1,372 @@
+        private final Semaphore finishedGetPermits = new Semaphore(0);
+
+        private WaitingSupplier() {
+        }
+
+        @Override
+        public Value apply(Key key) {
+            startedGetPermits.release(1);
+
+            try {
+                permits.acquire(1);
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+                fail();
+            }
+
+            finishedGetPermits.release(1);
+            return simpleTestValues.get(key.key());
+        }
+
+        /**
+         * Wait for a certain number of "gets" to have started. This will time out and fail the test after a certain amount of
+         * time if the "gets" never actually start.
+         */
+        public void waitForGetsToHaveStarted(int numExpectedGets) {
+            assertTrue(invokeSafely(() -> startedGetPermits.tryAcquire(numExpectedGets, 10, TimeUnit.SECONDS)));
+        }
+
+        /**
+         * Wait for a certain number of "gets" to have finished. This will time out and fail the test after a certain amount of
+         * time if the "gets" never finish.
+         */
+        public void waitForGetsToHaveFinished(int numExpectedGets) {
+            assertTrue(invokeSafely(() -> finishedGetPermits.tryAcquire(numExpectedGets, 10, TimeUnit.SECONDS)));
+        }
+
+        /**
+         * Release all threads blocked in this supplier.
+         */
+        @Override
+        public void close() {
+            permits.release(50);
+        }
+    }
+
+
+    private ExecutorService executorService;
+    private List<Future<?>> allExecutions;
+    private Supplier<LruCache<Key, Value>> simpleCache = () -> LruCache.builder(simpleValueSupplier)
+                                                                       .maxSize(MAX_SIMPLE_CACHE_SIZE)
+                                                                       .build();
+
+
+
+    @BeforeEach
+    public void setup() {
+        executorService = Executors.newFixedThreadPool(50);
+        allExecutions = new ArrayList<>();
+    }
+
+
+    @AfterEach
+    public void shutdown() {
+        executorService.shutdown();
+    }
+
+    @Test
+    void when_cacheHasMiss_ValueIsCalculatedAndCached() {
+        LruCache<Key, Value> cache = simpleCache.get();
+        populateAndVerify(cache, 1);
+    }
+
+    @Test
+    void when_cacheHasHit_ValueIsRetrievedFromCache() {
+        LruCache<Key, Value> cache = simpleCache.get();
+
+        populateAndVerify(cache, MAX_SIMPLE_CACHE_SIZE);
+        Value result = cache.get(simpleTestKeys.get(2));
+        assertThat(cache.size()).isEqualTo(MAX_SIMPLE_CACHE_SIZE);
+        assertThat(result).isNotNull();
+        assertThat(result).isEqualTo(simpleTestValues.get(2));
+
+        cache.get(simpleTestKeys.get(2));
+
+        int onlyOneInvocation = 1;
+        verify(simpleValueSupplier, times(onlyOneInvocation)).apply(simpleTestKeys.get(2));
+    }
+
+    @Test
+    void when_cacheFillsUp_ValuesAreEvictedFromCache() {
```

Comment:
It seems like this test is verifying 3 things: we evict when we're full, we promote things when they're used, and we don't call the underlying supplier more than needed. Should we split it into 3 tests? The name only seems to imply we are testing 1 of those things.

Reply:
Since there isn't an API that extracts the cache contents verification is indirect through checking calls to the supplier. It's hard to split the test strictly.